src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ArenaAllocator.java

```diff
@@ -63,22 +26,67 @@ private MemorySegment trySlice(long bytesSize, long bytesAlignment) {
         }
     }
 
-    private void checkConfinementIfNeeded() {
-        Thread segmentThread = segment.scope().ownerThread();
-        if (segmentThread != null && segmentThread != Thread.currentThread()) {
+    void checkConfinementIfNeeded() {
+        Thread ownerThread = scope().ownerThread();
+        if (ownerThread != null && ownerThread != Thread.currentThread()) {
             throw new IllegalStateException("Attempt to allocate outside confinement thread");
         }
     }
 
+    ResourceScope scope() {
+        return segment.scope();
+    }
+
+    public static class UnboundedArenaAllocator extends ArenaAllocator {
+
+        private static final long DEFAULT_BLOCK_SIZE = 4 * 1024;
+        private static final long DEFAULT_MAX_ALLOC_SIZE = DEFAULT_BLOCK_SIZE / 2;
+
+        public UnboundedArenaAllocator(ResourceScope scope) {
+            super(MemorySegment.allocateNative(DEFAULT_BLOCK_SIZE, 1, scope));
+        }
+
+        private MemorySegment newSegment(long size, long align) {
+            return MemorySegment.allocateNative(size, align, segment.scope());
+        }
+
+        @Override
+        public MemorySegment allocate(long bytesSize, long bytesAlignment) {
+            checkConfinementIfNeeded();
+            // try to slice from current segment first...
+            MemorySegment slice = trySlice(bytesSize, bytesAlignment);
+            if (slice != null) {
+                return slice;
+            } else {
+                if (Utils.alignUp(bytesSize, bytesAlignment) > DEFAULT_MAX_ALLOC_SIZE) {
```

Comment:
This seems to be the subject of some of the emails. I think maybe it might be worth it to break this out into an intermediate variable with a good name?

I also think the DEFAULT_MAX_ALLOC_SIZE could just be DEFAULT_BLOCK_SIZE, if we change this check to check for the maximum allocation size computed by `byteSize + bytesAlignment - 1` instead of using `alignUp`:

```suggestion
                long maxPossibleAllocationSize = bytesSize + bytesAlignment - 1;
                if (maxPossibleAllocationSize > DEFAULT_BLOCK_SIZE) {
```

e.g. if we have an allocation aligned to 8 bytes, we can have a maximum of 7 bytes of alignment padding (in case the address is off by just 1 byte).

Reply:
I've accepted this comment